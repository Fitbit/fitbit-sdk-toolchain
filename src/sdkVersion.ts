import semver from 'semver';

import packageVersionConst from './packageVersion.const';

export default function sdkVersion(toolchainVersion = packageVersionConst) {
  const version = semver.parse(toolchainVersion);
  if (version === null) {
    throw new Error(`Invalid SDK package version: ${toolchainVersion}`);
  }
  return version;
}

export function apiVersions(
  { enableProposedAPI }: { enableProposedAPI?: boolean } = {},
  toolchainVersion = packageVersionConst,
) {
  if (enableProposedAPI) return { deviceApi: '*', companionApi: '*' };

  const { major, minor } = sdkVersion(toolchainVersion);
  if (major === 1 && minor === 0) {
    return { deviceApi: '1.0.0', companionApi: '1.0.0' };
  }
  if (major === 2 && minor === 0) {
    return { deviceApi: '3.0.0', companionApi: '2.0.0' };
  }
  if (major === 3 && minor === 0) {
    return { deviceApi: '4.0.0', companionApi: '2.1.0' };
  }
  throw new Error(
    `No known API versions for SDK package version ${major}.${minor}`,
  );
}
