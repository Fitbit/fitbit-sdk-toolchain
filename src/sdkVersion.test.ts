import sdkVersion, { apiVersions } from './sdkVersion';

it('throws if package version has no known mapping', () => {
  expect(() => apiVersions({}, '1000.0.0')).toThrowErrorMatchingSnapshot();
});

it('provides a mapping for SDKv1', () => {
  expect(apiVersions({}, '1.0.0')).toEqual({
    deviceApi: '1.0.0',
    companionApi: '1.0.0',
  });
});

it('provides a mapping for SDKv2', () => {
  expect(apiVersions({}, '2.0.0')).toEqual({
    deviceApi: '3.0.0',
    companionApi: '2.0.0',
  });
});

it('provides a mapping for a known SDK version with a non-zero patch version', () => {
  expect(apiVersions({}, '1.0.100')).toEqual({
    deviceApi: '1.0.0',
    companionApi: '1.0.0',
  });
});

it('provides a mapping for a known SDK version with a pre-release suffix', () => {
  expect(apiVersions({}, '2.0.0-alpha.1')).toEqual({
    deviceApi: '3.0.0',
    companionApi: '2.0.0',
  });
});

it('maps API versions to "*" when proposed APIs are enabled', () => {
  expect(apiVersions({ enableProposedAPI: true }, '1000.0.0')).toEqual({
    deviceApi: '*',
    companionApi: '*',
  });
});

it('is able to derive the current SDK version', () => {
  expect(sdkVersion).not.toThrow();
});

it('is able to map the current SDK version to API versions', () => {
  expect(apiVersions()).toEqual({
    deviceApi: expect.any(String),
    companionApi: expect.any(String),
  });
});
