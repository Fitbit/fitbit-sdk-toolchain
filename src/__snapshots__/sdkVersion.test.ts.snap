// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if package version has no known mapping 1`] = `"No known API versions for SDK package version 1000.0"`;
